authorization/v2/uniformResourceIdentifier.go

```diff
@@ -157,3 +158,49 @@ func marshallURIMap(uriMap map[string]string) string {
 	// Remove the extra '&' character introduced when marshaling the uriMap
 	return marshalledMap[:len(marshalledMap)-1]
 }
+
+// isURIMatch checks that the target URI is a subset of the source URI (the URI which the user can access)
+func isURIMatch(source UniformResourceIdentifier, target UniformResourceIdentifier) bool {
+	// Ensure the target path is a subset of the source path
+	if len(source.path) > len(target.path) {
+		return false
+	}
+
+	// Compare URI path
+	for i, sourceCharacter := range source.path {
+		if uint8(sourceCharacter) != target.path[i] {
+			return false
+		}
+	}
+
+	// Validate URI arguments
+	for key, sourceValue := range source.arguments {
+		if targetValue, ok := target.arguments[key]; ok {
+			match, err := regexp.Match(sourceValue, []byte(targetValue))
+			if !match || err != nil {
+				// Fail-soft, if the regex is invalid or the regex pattern match fails, the URIs don't match
+				return false
+			}
+		}
+	}
+
+	// Validate URI metadata
+	for key, sourceValue := range source.metadata {
+		targetValue, ok := target.metadata[key]
+		if !ok || sourceValue != targetValue {
+			return false
+		}
+	}
+
+	return true
+}
+
+// CompareURI validates that the source URI (the URI for which the user has permissions) matches
+// one of the URIs in the target set
+func CompareURI(source UniformResourceIdentifier, targets []UniformResourceIdentifier) bool {
+	targetsMatch := true
+	for i := 0; targetsMatch && i < len(targets); i++ {
+		targetsMatch = isURIMatch(source, targets[i])
```

Comment:
I don't think that would do what we want. Say if you have may target URIs and only valid URI is in the last entry of the array. If we return false when the first URI we detect is false then it would never reach the valid URI.

Unless what we want is that the source URI matches **all** of the URIs in the target set?

Reply:
Right yeah, sorry I was thinking about how my proposed `isSubsetOfAtLeastOne` would work 😅 

With `isSubsetOfAtLeastOne` we could return `true` inside the loop and `false` at the end of the function.

Reply:
Though don't we already have the problem you described?

If `isURIMatch` returns `false`, `targetsMatch` gets set to `false`, the loop terminates immediately and the function returns `false`. It only returns true if `source` is a subset of all `targets`, right? Also it returns true if `targets` is empty which is not the required behaviour either.

Reply:
Oh true! I can fix that by flipping around the logic in the loop. Thanks for catching that!

Reply:
Could you also add a test for this scenario please 😊